dashdotdb/services/dora.py

```diff
@@ -0,0 +1,111 @@
+import logging
+import datetime
+from dataclasses import dataclass, field
+from typing import List
+
+from psycopg2.errors import UniqueViolation  # pylint: disable-msg=E0611
+from sqlalchemy import exc
+
+from dashdotdb.models.dashdotdb import db, Token, DataTypes, DORADeployment, DORACommit
+from dashdotdb.controllers.token import TokenNotFound
+
+
+@dataclass
+class DORAInsertStats:
+    created: List[str] = field(default_factory=lambda: [])
+    duplicated: List[str] = field(default_factory=lambda: [])
+    error: List[str] = field(default_factory=lambda: [])
+
+
+class DORA:
+    def __init__(self) -> None:
+        self.log = logging.getLogger()
+
+    def insert(self, token, manifest) -> DORAInsertStats:
+        db_token = (
+            db.session.query(Token)
+            .filter(Token.uuid == token, Token.data_type == DataTypes.DORADataType)
+            .first()
+        )
+
+        if db_token is None:
+            raise TokenNotFound()
+
+        stats = DORAInsertStats()
+
+        for dep_data in manifest["deployments"]:
+            # start transaction - we want to do an atomic
+            # transaction of deployment and commits
+            db.session.begin_nested()
+
+            dep_short_name = (
+                f"app_name={dep_data['app_name']},"
+                f"env_name={dep_data['env_name']},"
+                f"pipeline={dep_data['pipeline']},"
+                f"trigger_reason={dep_data['trigger_reason']}"
+            )
+
+            try:
+                finish_timestamp = datetime.datetime.fromisoformat(
+                    dep_data["finish_timestamp"]
+                )
+
+                # finish_timestamp and timestamp may come offset-naive or offset-aware,
+                # and they can't be subtracted if they are different. Converting always
+                # to offset-aware. If no TZ data is provided, we assume UTC.
+                if finish_timestamp.tzinfo is None:
+                    finish_timestamp = finish_timestamp.replace(
+                        tzinfo=datetime.timezone.utc
+                    )
+
+                deployment = DORADeployment(
+                    token_id=db_token.id,
+                    finish_timestamp=finish_timestamp,
+                    trigger_reason=dep_data["trigger_reason"],
+                    app_name=dep_data["app_name"],
+                    env_name=dep_data["env_name"],
+                    pipeline=dep_data["pipeline"],
+                )
+                db.session.add(deployment)
+                db.session.commit()
+
+                for commit_data in dep_data["commits"]:
```

Comment:
commits is not required in the schema, meaning it can be None here and break the for loop.

Reply:
fixed, added it as required in swagger. Can you confirm this is good? Should I also do `dep_data.get('commits', [])` or no need?

Reply:
nope, this is fine as is